compiler/rustc_mir_build/src/builder/scope.rs

```diff
@@ -887,7 +890,27 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     ),
                 }
             }
-            _ => self.as_constant(&self.thir[value]),
+
+            ExprKind::Literal { .. }
+            | ExprKind::NonHirLiteral { .. }
+            | ExprKind::ZstLiteral { .. }
+            | ExprKind::NamedConst { .. } => self.as_constant(&self.thir[value]),
```

Comment:
Generic constant items will have the same problem, as you could specify them with arguments.

Please add a test for them and for associated consts

Reply:
I'm not entirely sure what you mean, I added some extra test cases for the associated const scenarios I could think of. Did you have anything else in mind?

Reply:
Hmm, well the examples I can come up with all fail well before any const evaluation would actually take place. Anyway, I've added what I think you mean.